core/src/main/java/com/linecorp/armeria/client/retry/RetryStrategyBindingBuilder.java

```diff
@@ -0,0 +1,212 @@
+    }
+
+    /**
+     * Adds the specified {@link HttpStatusClass}es for a {@link RetryStrategy} which will retry
+     * if the class of the response status is one of the specified {@link HttpStatusClass}es.
+     */
+    public RetryStrategyBindingBuilder onStatusClass(Iterable<HttpStatusClass> statusClasses) {
+        requireNonNull(statusClasses, "statusClasses");
+        checkArgument(!Iterables.isEmpty(statusClasses), "statusClasses can't be empty");
+
+        statusClassesBuilder.addAll(statusClasses);
+        return this;
+    }
+
+    /**
+     * Adds the specified {@link HttpStatus}es for a {@link RetryStrategy} which will retry
+     * if the response status is one of the specified {@link HttpStatus}es.
+     */
+    public RetryStrategyBindingBuilder onStatus(HttpStatus... statuses) {
+        return onStatus(ImmutableSet.copyOf(requireNonNull(statuses, "statuses")));
+    }
+
+    /**
+     * Adds the specified {@link HttpStatus}es for a {@link RetryStrategy} which will retry
+     * if the response status is one of the specified {@link HttpStatus}es.
+     */
+    public RetryStrategyBindingBuilder onStatus(Iterable<HttpStatus> statuses) {
+        requireNonNull(statuses, "statuses");
+        checkArgument(!Iterables.isEmpty(statuses), "statuses can't be empty");
+
+        statusesBuilder.addAll(statuses);
+        return this;
+    }
+
+    /**
+     * Adds the specified exception type for a {@link RetryStrategy} which will retry
+     * if an {@link Exception} is raised and that is instance of the specified {@code exception}.
+     */
+    public RetryStrategyBindingBuilder onException(Class<? extends Throwable> exception) {
+        requireNonNull(exception, "exception");
+        return onException(exception::isInstance);
+    }
+
+    /**
+     * Adds the specified {@code exceptionFilter} for a {@link RetryStrategy} which will retry
+     * if an {@link Exception} is raised and the specified {@code exceptionFilter} returns {@code true}.
+     */
+    public RetryStrategyBindingBuilder onException(Predicate<? super Throwable> exceptionFilter) {
+        requireNonNull(exceptionFilter, "exceptionFilter");
+        if (this.exceptionFilter != null) {
+            this.exceptionFilter = this.exceptionFilter.or(exceptionFilter);
+        } else {
+            @SuppressWarnings("unchecked")
+            final Predicate<Throwable> cast = (Predicate<Throwable>) exceptionFilter;
+            this.exceptionFilter = cast;
+        }
+        return this;
+    }
+
+    /**
+     * Sets the specified {@link Backoff} and returns the {@link RetryStrategyBuilder} that this
+     * {@link RetryStrategyBindingBuilder} was created from.
+     *
+     * @throws IllegalStateException if none of {@link HttpStatus}, {@link HttpStatusClass} or
+     *                               an expected {@code exception} type is specified
+     */
+    public RetryStrategyBuilder build(Backoff backoff) {
```

Comment:
I think a user will want to specify different backoff for different situation, rather than using a single `Backoff`.

How about using a fluent API like this: (Just a draft)

```java
RetryStrategy.builder()
             .on().idempotentMethod()
                    .then(someBackoff)
             .on()....then(otherBackoff)
             .on()....thenStop()
             .on()....thenImmediately()
             .build();
```

If no conditions match, could skip retrying.

Reply:
> I think a user will want to specify different backoff for different situation

Oh. That sounds good. Let me improve the API 👍

Reply:
Or could accept a customizer?
```java
RetryStrategy.builder()
             .on(retryOnError ->
                 retryOnError.idempotentMethods()
                             .exception(ResponseTimeoutException.class)
                             .then(Backoff))
             .on(noRetry -> 
                 noRetry.status(HttpStatus.BAD_REQUEST)
                        .thenStop())
             .on(retryOnStatus ->
                 retryOnStatus.status(HttpStatus.INTERNAL_SERVER_ERROR)
                              .exception(ex -> ex instanceof ClosedSessionException)
                              .method(HttpMethod.GET)
                              .thenNoDelay())
             .build()
```